### ISSUE TITLE
Gpush does not return to vim

### ISSUE THREAD

**Issue by fedemp** [tpope/vim-fugitive#531]
Hi, Tim. I have just update fugitive. Used `:Gpush`, vim goes to background and I see the typical git messages about my commits being pushed. And then nothing more. The "Press RETURN to return to vim" does not appear. I have to `^C` to go back to vim.

**Comment by languitar** [comment]
I am experiencing the same issue and I can tell for sure that git push returns correctly all the time.

**Comment by languitar** [comment]
Actually, inside neovim the problem is fatal because there is no way to recover the vim instance without killing it.

I sometimes had the feeling that this might be related to the use of an external credential helper used by git. Can anyone confirm this?

**Comment by languitar** [comment]
Ok, I finally got a clue what is happening here. So I request that this issue is opened again.

It seems that this is related to the credential caching mechanism available in git. In my `$HOME/.gitconfig` I have got the following fragment:
```
[credential]
        helper = cache
```
With this setting, git sometimes spawns a second process called `git-credential-cache--daemon`:
```
jwienke  16551     1  0 14:28 pts/12   00:00:00 git-credential-cache--daemon /homes/jwienke/.git-credential-cache/socket
```
As visible in the ps output, the git process that created this helper process (which is the process performing the push operation) handed this process over to init so that it becomes a daemon. However, the `Gpush` operation also seems to wait on this process. The hanging vim immediately returned once I killed the daemon process.

**Comment by bmatcuk** [comment]
I know this issue is old, but I'm running into it as well and this is the only place on the entire internet where it is mentioned (other than the [stackoverflow](http://stackoverflow.com/questions/30745536/vi-fugitive-gpush-does-not-exit-sometimes) question I opened), so I thought I'd add my findings...

I'm not using any kind of credential helper like @languitar above, but I have setup ssh to persist connections. With the following in my ~/.ssh/config:

```
Host *
    ControlPersist 5m
```

...when I do a `:Gpush`, ssh spawns off a process to keep the connection alive. This triggers the issue. I'm using tmux and vim-dispatch, so, in my case, I get a tmux pane that stays open. If I do a second push, while the connection is still alive, I have no issues (a pane opens, the push happens, the pane closes). If I wait 5 minutes, the ssh process exits and the pane closes. Oddly, despite the fact that the pane is waiting on the ssh process, if I switch to that pane and Ctrl+C, it's not killing the ssh process... so I'm not sure what process is actually hanging. A `ps` shows the git process, itself, exits fine.

So, tl;dr: the issue is triggered whenever git spawns off other processes. In my case, it's ssh... in languitar's case, it was git-credential-cache-daemon... not sure about OP, but my guess is he had something similar configured as well.

One possible solution I've come up with for my particular case is to disable the ControlPersist option, but only for git commands run inside vim via fugitive. To do this, I've added the following to my .vimrc:

```
let g:fugitive_git_executable = "env GIT_SSH_COMMAND='ssh -o ControlPersist=no' git"
```

It's ugly and makes me sad that I have to disable the persistent connection, but it works...

**Comment by tpope** [comment]
FWIW I ended up doing away with ControlPersist in my ssh config long ago due to similar issues, though nothing Vim specific. Perhaps it has to do with the parent shell? If you `git push` from the shell and then immediately `logout`, does that hang too?

**Comment by bmatcuk** [comment]
I don't seem to have a problem with that... I made sure ssh was dead and then I ran `git push` followed immediately by `logout`... worked fine.

*Some other things I tried:*

I tried altering [fugitive.vim, line 1586](https://github.com/tpope/vim-fugitive/blob/master/plugin/fugitive.vim#L1586) to run makeprg via `bash -c`, to see if that might help, but that didn't change anything.

My next thought was to have it run git in the background (with `&`) and then `disown $!` and `wait $!` by editing that same line in fugitive.vim, but something in vim-dispatch blew up on the `&`. So I moved that stuff into a shell script and had fugitive call it by pointing `g:fugitive_git_executable` to the script, but that didn't change anything either =(

I'm not super sad about disabling ControlPersist for pushes and pulls since I tend to do them much less often than 5 minutes (my ControlPersist setting) anyway... it's just that the problem solver in me is kind of curious what the problem is =) Oh well...

Thanks for the great plugins!

**Comment by languitar** [comment]
Maybe we can at least get around the caching daemon problem by passing the option to disable it to the git call via a `-c` flag?

**Comment by tpope** [comment]
If you can confirm it helps and send a pull request, I will merge.
